File: src/core/deployment/strategy.ts
import { invalidArgument } from "../errors";
import type {
  DeployProvider,
  DeploymentProviderCandidate,
  DeploymentProviderPolicy,
  DeploymentCodeProbe,
  DeploymentSourceModel,
} from "./types";
import type { ExistingDeploymentFiles } from "./existing";

export type DeploymentStrategy = {
  provider: DeployProvider;
  reason: string;
  policy: DeploymentProviderPolicy;
  candidates: DeploymentProviderCandidate[];
};

export function resolveDeploymentStrategy(input: {
  codeProbe: DeploymentCodeProbe;
  existing: ExistingDeploymentFiles;
  policy?: Partial<DeploymentProviderPolicy>;
}): DeploymentStrategy {
  const policy = normalizeProviderPolicy(input.policy);
  validateProviderPolicy({ ...input, policy });
  const selectedProvider = selectProvider({ ...input, policy });

  return {
    provider: selectedProvider,
    reason: reasonFor(selectedProvider, { ...input, policy }),
    policy,
    candidates: providerCandidates(selectedProvider, { ...input, policy }),
  };
}

export function deploymentStrategyForSourceModel(input: {
  strategy: DeploymentStrategy;
  sourceModel: DeploymentSourceModel;
}): DeploymentStrategy {
  if (input.sourceModel.services.length <= 1 || input.strategy.provider !== "dockerfile-existing") {
    return input.strategy;
  }

  return {
    ...input.strategy,
    provider: "dockerfile-template",
    reason: "Deployment source model has multiple application services, so loom will generate service-specific Dockerfile/Compose assets instead of reusing one root Dockerfile as the whole application.",
    candidates: input.strategy.candidates.map((candidate) => {
      if (candidate.provider === "dockerfile-existing") {
        return {
          ...candidate,
          status: "skipped" as const,
          reason: "Skipped because one root Dockerfile cannot represent multiple application services from the deployment source model.",
        };
      }
      if (candidate.provider === "dockerfile-template") {
        return {
          ...candidate,
          status: "selected" as const,
          reason: "Selected because the deployment source model requires service-specific generated Dockerfiles.",
        };
      }
      return candidate;
    }),
  };
}

function validateProviderPolicy(input: {
  existing: ExistingDeploymentFiles;
  policy: DeploymentProviderPolicy;
}): void {
  if (input.policy.forceGenerate && input.policy.provider && input.policy.provider !== "dockerfile-template") {
    throw invalidArgument("--force-generate cannot be combined with an existing-asset provider.", {
      provider: input.policy.provider,
    });
  }
  if (input.policy.provider === "compose-existing" && !input.existing.composePath) {
    throw invalidArgument("Provider policy selected compose-existing, but no root-level Compose file was found.");
  }
  if (input.policy.provider === "dockerfile-existing" && !input.existing.dockerfilePath) {
    throw invalidArgument("Provider policy selected dockerfile-existing, but no root-level Dockerfile was found.");
  }
}

function selectProvider(input: {
  codeProbe: DeploymentCodeProbe;
  existing: ExistingDeploymentFiles;
  policy: DeploymentProviderPolicy;
}): DeployProvider {
  if (input.policy.forceGenerate) {
    return "dockerfile-template";
  }
  if (input.policy.provider) {
    return input.policy.provider;
  }
  if (input.policy.reuseExisting && input.existing.composePath) {
    return "compose-existing";
  }
  if (input.policy.reuseExisting && input.existing.dockerfilePath) {
    return "dockerfile-existing";
  }
  return "dockerfile-template";
}

function reasonFor(
  provider: DeployProvider,
  input: {
    codeProbe: DeploymentCodeProbe;
    existing: ExistingDeploymentFiles;
    policy: DeploymentProviderPolicy;
  },
): string {
  if (input.policy.forceGenerate) {
    return "Provider policy forces generated Dockerfile/Compose assets.";
  }
  if (input.policy.provider) {
    return `Provider policy explicitly selected ${input.policy.provider}.`;
  }
  if (!input.policy.reuseExisting) {
    return "Provider policy disables existing deployment asset reuse, so loom will generate Dockerfile/Compose assets.";
  }

  switch (provider) {
    case "compose-existing":
      return "Root-level Compose file exists, so loom will reuse it without overwriting user deployment assets.";
    case "dockerfile-existing":
      return "Root-level Dockerfile exists, so loom will reuse it and materialize only a local Compose wrapper.";
    case "dockerfile-template":
      return `Repository probes found ${input.codeProbe.kind} runtime evidence, so loom will generate deployment files from the deployment source model.`;
  }
}

function providerCandidates(
  selectedProvider: DeployProvider,
  input: {
    codeProbe: DeploymentCodeProbe;
    existing: ExistingDeploymentFiles;
    policy: DeploymentProviderPolicy;
  },
): DeploymentProviderCandidate[] {
  return [
    {
      provider: "compose-existing",
      status: statusFor("compose-existing", selectedProvider, Boolean(input.existing.composePath)),
      reason: candidateReason("compose-existing", input),
      commands: [["docker", "compose", "config", "--quiet"]],
    },
    {
      provider: "dockerfile-existing",
      status: statusFor("dockerfile-existing", selectedProvider, Boolean(input.existing.dockerfilePath)),
      reason: candidateReason("dockerfile-existing", input),
      commands: [["docker", "compose", "up", "-d", "--build"]],
    },
    {
      provider: "dockerfile-template",
      status: selectedProvider === "dockerfile-template" ? "selected" : "available",
      reason: candidateReason("dockerfile-template", input),
      commands: [["docker", "compose", "up", "-d", "--build"]],
    },
  ];
}

function candidateReason(
  provider: DeployProvider,
  input: {
    codeProbe: DeploymentCodeProbe;
    existing: ExistingDeploymentFiles;
    policy: DeploymentProviderPolicy;
  },
): string {
  if (input.policy.forceGenerate && provider !== "dockerfile-template") {
    return "Skipped because provider policy forces generated Dockerfile/Compose assets.";
  }
  if (input.policy.provider && input.policy.provider !== provider) {
    return `Skipped because provider policy explicitly selected ${input.policy.provider}.`;
  }
  if (!input.policy.reuseExisting && provider !== "dockerfile-template") {
    return "Skipped because provider policy disables existing deployment asset reuse.";
  }

  switch (provider) {
    case "compose-existing":
      return input.existing.composePath
        ? "Existing Compose file found at project root."
        : "No root-level Compose file was found.";
    case "dockerfile-existing":
      return input.existing.dockerfilePath
        ? "Existing Dockerfile found at project root."
        : "No root-level Dockerfile was found.";
    case "dockerfile-template":
      return templateReason(input.codeProbe.kind);
  }
}

function statusFor(
  provider: DeployProvider,
  selectedProvider: DeployProvider,
  isAvailable: boolean,
): DeploymentProviderCandidate["status"] {
  if (provider === selectedProvider) {
    return "selected";
  }
  return isAvailable ? "available" : "skipped";
}

export function normalizeProviderPolicy(
  policy?: Partial<DeploymentProviderPolicy>,
): DeploymentProviderPolicy {
  return {
    provider: policy?.provider,
    reuseExisting: policy?.forceGenerate ? false : policy?.reuseExisting ?? true,
    forceGenerate: policy?.forceGenerate ?? false,
  };
}

function templateReason(kind: DeploymentCodeProbe["kind"]): string {
  if (kind === "unknown") {
    return "Available as a deterministic local preview so a coding agent can inspect and repair deployment files if the first attempt fails.";
  }
  return `Available because loom can model a ${kind} runtime as a deployment source service.`;
}

File: src/rust/deploy/strategy.rs
use contracts::{
    DeployProvider, DeploymentProviderCandidate, DeploymentProviderCandidateStatus,
    DeploymentProviderPolicy, DeploymentSourceModel,
};

use crate::{code_evidence::DeploymentCodeProbe, existing::ExistingDeploymentFiles};

#[derive(Debug, Clone)]
pub struct DeploymentStrategy {
    pub provider: DeployProvider,
    pub reason: String,
    pub policy: DeploymentProviderPolicy,
    pub candidates: Vec<DeploymentProviderCandidate>,
}

pub fn resolve_deployment_strategy(
    code_probe: &DeploymentCodeProbe,
    source_model: &DeploymentSourceModel,
    existing: &ExistingDeploymentFiles,
    policy: Option<DeploymentProviderPolicy>,
) -> DeploymentStrategy {
    let policy = normalize_provider_policy(policy);
    let provider = select_provider(source_model, existing, &policy);
    let reason = reason_for(provider, code_probe, source_model, existing, &policy);
    let candidates = provider_candidates(provider, code_probe, source_model, existing, &policy);
    DeploymentStrategy {
        provider,
        reason,
        policy,
        candidates,
    }
}

pub fn normalize_provider_policy(
    policy: Option<DeploymentProviderPolicy>,
) -> DeploymentProviderPolicy {
    let Some(mut policy) = policy else {
        return DeploymentProviderPolicy::default();
    };
    if policy.force_generate {
        policy.provider = Some(DeployProvider::Generated);
        policy.reuse_existing = false;
    }
    policy
}

fn select_provider(
    source_model: &DeploymentSourceModel,
    existing: &ExistingDeploymentFiles,
    policy: &DeploymentProviderPolicy,
) -> DeployProvider {
    if policy.force_generate {
        return DeployProvider::Generated;
    }
    if let Some(provider) = policy.provider {
        return provider;
    }
    if !policy.reuse_existing {
        return DeployProvider::Generated;
    }
    if existing.compose_path.is_some() {
        return DeployProvider::ComposeExisting;
    }
    if existing.dockerfile_path.is_some() && source_model.services.len() <= 1 {
        return DeployProvider::DockerfileExisting;
    }
    DeployProvider::Generated
}

fn reason_for(
    provider: DeployProvider,
    code_probe: &DeploymentCodeProbe,
    source_model: &DeploymentSourceModel,
    existing: &ExistingDeploymentFiles,
    policy: &DeploymentProviderPolicy,
) -> String {
    if policy.force_generate {
        return "Provider policy forces generated Dockerfile/Compose assets.".to_string();
    }
    if let Some(forced) = policy.provider {
        return format!(
            "Provider policy explicitly selected {}.",
            provider_label(forced)
        );
    }
    if !policy.reuse_existing {
        return "Provider policy disables existing deployment asset reuse.".to_string();
    }
    match provider {
        DeployProvider::ComposeExisting => {
            "Root-level Compose file exists, so Loom will try it before generated fallback."
                .to_string()
        }
        DeployProvider::DockerfileExisting => {
            "Root-level Dockerfile exists, so Loom will reuse it with a generated Compose wrapper."
                .to_string()
        }
        DeployProvider::Generated => {
            if existing.dockerfile_path.is_some() && source_model.services.len() > 1 {
                "Existing root Dockerfile cannot represent multiple application services; generated deployment assets are safer.".to_string()
            } else {
                format!(
                    "Repository probes found {:?} runtime evidence, so Loom will generate deployment assets.",
                    code_probe.kind
                )
            }
        }
    }
}

fn provider_candidates(
    selected: DeployProvider,
    code_probe: &DeploymentCodeProbe,
    source_model: &DeploymentSourceModel,
    existing: &ExistingDeploymentFiles,
    policy: &DeploymentProviderPolicy,
) -> Vec<DeploymentProviderCandidate> {
    [
        DeployProvider::ComposeExisting,
        DeployProvider::DockerfileExisting,
        DeployProvider::Generated,
    ]
    .into_iter()
    .map(|provider| DeploymentProviderCandidate {
        provider,
        status: candidate_status(provider, selected, source_model, existing, policy),
        reason: candidate_reason(provider, code_probe, source_model, existing, policy),
        commands: candidate_commands(provider),
    })
    .collect()
}

fn candidate_status(
    provider: DeployProvider,
    selected: DeployProvider,
    source_model: &DeploymentSourceModel,
    existing: &ExistingDeploymentFiles,
    policy: &DeploymentProviderPolicy,
) -> DeploymentProviderCandidateStatus {
    if provider == selected {
        return DeploymentProviderCandidateStatus::Selected;
    }
    if policy.force_generate && provider != DeployProvider::Generated {
        return DeploymentProviderCandidateStatus::Skipped;
    }
    if policy.provider.is_some_and(|forced| forced != provider) {
        return DeploymentProviderCandidateStatus::Skipped;
    }
    match provider {
        DeployProvider::ComposeExisting => existing
            .compose_path
            .as_ref()
            .map(|_| DeploymentProviderCandidateStatus::Available)
            .unwrap_or(DeploymentProviderCandidateStatus::Skipped),
        DeployProvider::DockerfileExisting => {
            if existing.dockerfile_path.is_none() || source_model.services.len() > 1 {
                DeploymentProviderCandidateStatus::Skipped
            } else {
                DeploymentProviderCandidateStatus::Available
            }
        }
        DeployProvider::Generated => DeploymentProviderCandidateStatus::Available,
    }
}

fn candidate_reason(
    provider: DeployProvider,
    code_probe: &DeploymentCodeProbe,
    source_model: &DeploymentSourceModel,
    existing: &ExistingDeploymentFiles,
    policy: &DeploymentProviderPolicy,
) -> String {
    if policy.force_generate && provider != DeployProvider::Generated {
        return "Skipped because provider policy forces generated deployment assets.".to_string();
    }
    if let Some(forced) = policy.provider {
        if forced != provider {
            return format!(
                "Skipped because provider policy explicitly selected {}.",
                provider_label(forced)
            );
        }
    }
    match provider {
        DeployProvider::ComposeExisting => existing
            .compose_path
            .as_ref()
            .map(|_| "Existing Compose file found at deployment root.".to_string())
            .unwrap_or_else(|| "No root-level Compose file was found.".to_string()),
        DeployProvider::DockerfileExisting => {
            if existing.dockerfile_path.is_none() {
                "No root-level Dockerfile was found.".to_string()
            } else if source_model.services.len() > 1 {
                "Skipped because one root Dockerfile cannot represent multiple application services.".to_string()
            } else {
                "Existing Dockerfile found at deployment root.".to_string()
            }
        }
        DeployProvider::Generated => format!(
            "Available because Loom can model {:?} runtime evidence as generated local deployment assets.",
            code_probe.kind
        ),
    }
}

fn candidate_commands(provider: DeployProvider) -> Vec<Vec<String>> {
    match provider {
        DeployProvider::ComposeExisting => {
            vec![vec![
                "docker".to_string(),
                "compose".to_string(),
                "config".to_string(),
                "--quiet".to_string(),
            ]]
        }
        DeployProvider::DockerfileExisting | DeployProvider::Generated => vec![vec![
            "docker".to_string(),
            "compose".to_string(),
            "up".to_string(),
            "-d".to_string(),
            "--build".to_string(),
        ]],
    }
}

fn provider_label(provider: DeployProvider) -> &'static str {
    match provider {
        DeployProvider::ComposeExisting => "compose-existing",
        DeployProvider::DockerfileExisting => "dockerfile-existing",
        DeployProvider::Generated => "generated",
    }
}
